### ISSUE TITLE
CA1816 (Call SuppressFinalize) false positive when not using the Dispose pattern

### ISSUE THREAD

**Issue by sharwell** [dotnet/roslyn-analyzers#3925]
#### Analyzer package

Unknown

#### Package Version

3.3.0-beta1.20352.4

#### Diagnostic ID

[CA1816](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1816)

#### Repro steps

```csharp
public class MyType : IDisposable
{
  public void Dispose() { }
}
```

#### Expected behavior

CA1816 is not reported because `MyType` does not follow the dispose pattern.

#### Actual behavior

CA1816 is reported.

#### Additional details

I believe the specific conditions CA1816 should look for are the following:

1. The type or any of its base types has a user-defined finalizer (i.e. overrides `object.Finalize()`)
2. The type has both a `Dispose()` and a `Dispose(bool)` method

**Comment by sharwell** [comment]
Superseded by #4801
